blobby/src/lib.rs

```diff
@@ -114,6 +114,91 @@ fn read_vlq(data: &[u8], pos: &mut usize) -> Result<usize, Error> {
     Ok(val)
 }
 
+/// Write a git-flavoured VLQ value into `buf`.
+///
+/// Returns the slice within `buf` that holds the value.
+fn encode_vlq(mut val: usize, buf: &mut [u8; 4]) -> &[u8] {
+    macro_rules! step {
+        ($n:expr) => {
+            buf[$n] = if $n == 3 {
+                (val & (VAL_MASK as usize)) as u8
+            } else {
+                val -= 1;
+                NEXT_MASK | (val & (VAL_MASK as usize)) as u8
+            };
+            val >>= 7;
+            if val == 0 {
+                return &buf[$n..];
+            }
+        };
+    }
+
+    step!(3);
+    step!(2);
+    step!(1);
+    step!(0);
+    panic!("integer is too big")
+}
+
+/// Encode the given collection of binary blobs in .blb format into `writer`.
+/// Returns a count of the number of blobs included in the index.
+///
+/// The encoded file format is:
+///  - count of index entries=N
+///  - N x index entries, each encoded as:
+///      - size L of index entry (VLQ)
+///      - index blob contents (L bytes)
+///  - repeating encoded blobs, each encoded as:
+///      - VLQ value that is either:
+///         - (J << 1) & 0x01: indicates this blob is index entry J
+///         - (L << 1) & 0x00: indicates an explicit blob of len L
+///      - (in the latter case) explicit blob contents (L bytes)
+pub fn encode_blobs(blobs: Vec<Vec<u8>>, out_buf: &mut Vec<u8>) -> usize {
```

Comment:
> Note that we don't need for this function to return number of processed records

It's returning the number of entries in the index/dictionary rather than the number of input blobs, mostly so that the output of the `convert` tool would stay identical (it prints `Index len: 60`).

Reply:
I've changed to return a `Vec<u8>` but not changed the input parameter types (the trait bounds need some combination of `ExactSizeIterator` that I couldn't quite figure out).

Reply:
The signature which I've provided in the snippet [works without issues](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e645d8ac7c325cf24668b5fe1423380a) (with a small tweak).

>It's returning the number of entries in the index/dictionary rather than the number of input blobs, mostly so that the output of the convert tool would stay identical

Ah, I forgot about that. I think we can simply remove the print part, I was using it mostly for debug purposes.

Reply:
> The signature which I've provided in the snippet works without issues (with a small tweak).

Ah, thanks - I missed a place that needed an `.as_ref()` and the compiler sent me down a blind alley with `ExactSizeIterator` errors &hellip;